cmd/apiserver-watcher/gcp_handler.go

```diff
@@ -0,0 +1,215 @@
+//
+// Example rules using vips 10.0.0.2, 35.184.123.49 and 35.188.173.243
+//
+// -A PREROUTING -m comment --comment "gcp LB vip DNAT" -j gcp-vips
+// -A OUTPUT -m comment --comment "gcp LB vip DNAT for local clients" -j gcp-vips
+// -A gcp-vips -d 10.0.0.2/32 -j REDIRECT
+// -A gcp-vips -d 35.184.123.49/32 -j REDIRECT
+// -A gcp-vips -d 35.188.173.243/32 -j REDIRECT
+// -A gcp-vips-local -d 35.184.123.49/32 -j REDIRECT
+// -A gcp-vips-local -d 35.188.173.243/32 -j REDIRECT
+// -A gcp-vips-local -d 10.0.0.2/32 -j REDIRECT
+
+const (
+	// FIXME: support multiple network interfaces
+	gcpMetadataForwardedIPs = "instance/network-interfaces/0/forwarded-ips/?recursive=true"
+	// name of nat chain for iptables masquerade rules
+	gcpMasqChainName      = "gcp-vips"
+	gcpLocalMasqChainName = "gcp-vips-local"
+)
+
+type gcpHandler struct {
+	vip string
+	// GCP Loadbalancers provide the forwarded IPs in the metadata instance
+	forwardedIPs []string
+	ipt          *iptables.IPTables
+	stopCh       chan struct{}
+}
+
+var _ handler = &gcpHandler{}
+
+func newGCPHandler(vip string) (handler, error) {
+	ipt, err := iptables.New(
+		iptables.IPFamily(iptables.ProtocolIPv4),
+		iptables.Timeout(iptablesTimeout),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return &gcpHandler{
+		vip:    vip,
+		ipt:    ipt,
+		stopCh: make(chan struct{}),
+	}, nil
+}
+
+// onFailure: stop syncing iptables rules but don't flush all of them
+// keep the ones that are still being forwarded to this host.
+func (gcp *gcpHandler) onFailure() error {
+	select {
+	case gcp.stopCh <- struct{}{}:
+		glog.V(4).Info("stopping iptables sync loop")
+	default:
+		glog.V(4).Info("no sync iptables running")
+	}
+	// wait for iptables sync loop to finish
+	time.Sleep(iptablesTimeout * time.Second)
+
+	glog.Info("apiserver down: deleting iptables apiserver rules for local traffic")
+	return gcp.ipt.DeleteIfExists("nat", gcpLocalMasqChainName, "--dst", gcp.vip, "-j", "REDIRECT")
+}
+
+// onSuccess: either start routes service, or remove down file
+func (gcp *gcpHandler) onSuccess() error {
+	if err := gcp.syncRulesOnce(); err != nil {
+		return err
+	}
+
+	ip := net.ParseIP(gcp.vip)
+	// bz1930457 delete stale conntrack entries to the apiserver
+	filter := &netlink.ConntrackFilter{}
+	filter.AddIP(netlink.ConntrackNatSrcIP, ip)
+	if _, err := netlink.ConntrackDeleteFilter(netlink.ConntrackTable, netlink.FAMILY_V4, filter); err != nil {
+		glog.V(4).Infof("Error deleting conntrack entries for %s", gcp.vip)
+	}
+
+	go gcp.syncRulesUntil()
+	return nil
+
+}
+
+// https://github.com/openshift/installer/blob/master/data/data/gcp/network/lb-private.tf
+// healthy_threshold   = 3
+// unhealthy_threshold = 3
+// check_interval_sec  = 2
+// timeout_sec         = 2
+
+func (gcp *gcpHandler) successThreshold() int {
+	return 1
+}
+
+func (gcp *gcpHandler) failureThreshold() int {
+	// LB = 6 seconds, plus 10 seconds for propagation
+	return 8
+}
+
+// syncRulesOnce syncs ip masquerade rules
+// TODO ipv6
```

Comment:
is there a reason this is a TODO and not done?

Reply:
gcp doesn't support IPv6 yet, I can remove the comment or explain the reason, whatever you think is better

Reply:
explain is cool with me :)